### PR TITLE
Add HuggingFace Model Wrapper for Multi-Class Classification

### PR DESCRIPTION
**Pull request by harimkang**
### Summary

- Add HuggingFaceModelForMulticlassCls
- Training, Testing, Exporting works

image
*The results of running the existing models are from workstation, and for the hugging face model, the results are from a different machine, so it is difficult to compare the time directly. Just a note.

### Checked models
- "nateraw/vit-base-beans", # Available
- "google/vit-base-patch16-224", # Available
- "google/mobilenet_v1_1.0_224", # Available
- "google/efficientnet-b0", # Available
- "microsoft/beit-base-patch16-224-pt22k-ft22k", # Available
- "microsoft/swin-tiny-patch4-window7-224", # Available
- "microsoft/resnet-18", # Available **(but show NAN loss)**
- "facebook/convnextv2-tiny-1k-224", # Available
- "facebook/dinov2-small-imagenet1k-1-layer", # Available
- "facebook/deit-tiny-patch16-224", # Available

I haven't checked them all (too many models), but most of the models in the list below are available.
https://huggingface.co/models?pipeline_tag=image-classification

image

### How to test

Install transformers
`pip install transformers`

API Examples
```python
from otx.algo.classification.huggingface_model import HuggingFaceModelForMulticlassCls
from otx.engine import Engine

data_root = "/datasets/otx_v2_dataset/multiclass_classification/multiclass_food101_large"

otx_model = HuggingFaceModelForMulticlassCls(
    model_name_or_path="facebook/deit-tiny-patch16-224",
    label_info=20,
)

engine = Engine(
    data_root=data_root,
    model=otx_model,
    work_dir="otx-workspace-hf",
)

engine.train(max_epochs=2)
engine.test()
engine.export()
engine.test(checkpoint="otx-workspace-hf/exported_model.xml")
```

CLI Examples
```shell
(venv) ~/otx$ otx train \
    --model otx.algo.classification.huggingface_model.HuggingFaceModelForMulticlassCls \
    --model.model_name_or_path facebook/deit-tiny-patch16-224 \
    --data_root /datasets/otx_v2_dataset/multiclass_classification/multiclass_food101_large \
    --work_dir otx-workspace-hf-cli --max_epochs 3
```

### Checklist

- [ ] I have added unit tests to cover my changes.​
- [ ] I have added integration tests to cover my changes.​
- [ ] I have ran e2e tests and there is no issues.
- [ ] I have added the description of my changes into CHANGELOG in my target branch (e.g., [CHANGELOG](https://github.com/openvinotoolkit/training_extensions/blob/develop/CHANGELOG.md) in develop).​
- [ ] I have updated the documentation in my target branch accordingly (e.g., [documentation](https://github.com/openvinotoolkit/training_extensions/tree/develop/docs) in develop).
- [ ] I have [linked related issues](https://help.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword).

### License

- [ ] I submit _my code changes_ under the same [Apache License](https://github.com/openvinotoolkit/training_extensions/blob/develop/LICENSE) that covers the project.
      Feel free to contact the maintainers if that's a concern.
- [ ] I have updated the license header for each file (see an example below).

```python
# Copyright (C) 2024 Intel Corporation
# SPDX-License-Identifier: Apache-2.0
```

### CODE REVIEW

File: pyproject.toml [hunk 1]
```diff
@@ -103,6 +103,10 @@ base = [
     "anomalib[core]==1.1.0",
 ]
 
+transformers = [
```
Comment (eunwoosh) [inline pyproject.toml:106]:
```suggestion
huggingface = [
```
It's for enabling huggingface, so how about to change like this?

**Review by eunwoosh — COMMENTED** [review]
Thanks for your work :) I left minor comment, Please take a look.

File: pyproject.toml (hunk 1 above)
Comment (harimkang) [reply pyproject.toml:106]: This is ambiguous, but the library is actually a `transformers` integration. It's an integration of the transformers api, which supports the hugging face model, so I think transformers makes more sense for me. What do you think?

Comment (eunwoosh) [reply pyproject.toml:106]: I see. my thought is that the goal of integrating a `transformers` is for using huggingface model and name `transformers` can mean various thing. So, I think when user wants to use huggingface model, using `pip install otx[huggingface]` is more intuitive. This is just my thought. If you think different please resolve this :)

**Review by eunwoosh — APPROVED** [review]

File: src/otx/algo/classification/huggingface_model.py
Comment (sungchul1) [inline src/otx/algo/classification/huggingface_model.py:1]: Of course mmx will be removed, but what do you think about that gathering 3rd party frameworks' (huggingface and torchvision) models into `otx/core/model/classification.py`?
Like
https://github.com/openvinotoolkit/training_extensions/blob/e67a5fcd17fb350191771ed79dd8d4ed246c0b10/src/otx/core/model/classification.py#L90-L233

Comment (sungchul1) [reply src/otx/algo/classification/huggingface_model.py:1]: Actually, there are usually model-related files in task-specific directory:
image
I think supporting various frameworks is optional, isn't it?
If no, feel free to ignore it :D

Comment (harimkang) [reply src/otx/algo/classification/huggingface_model.py:1]: I'm not sure, I think `core` should be a parent class that can be used as a base, and `algo` will be the actual model implementation. (As a user, it seems appropriate to use the `algo`'s model class to build the model.) Also, `transformers` should be installed as optional when I discussed with Kirill, it is not installed in the current test env (not `base`). So I still think this is better off in an `algo`. @kprokofi what do you think?

File: pyproject.toml (hunk 1 above)
Comment (harimkang) [reply pyproject.toml:106]: Yeah we'll have to think about this a bit more, thanks for the comment!

**Review by sungchul1 — APPROVED** [review]

File: src/otx/algo/classification/huggingface_model.py
Comment (kprokofi) [reply src/otx/algo/classification/huggingface_model.py:1]: Yeah, I agree with Harim, we should use core/model as the base. I also thinking about removing MMmodels from core.

**Pull request merged**